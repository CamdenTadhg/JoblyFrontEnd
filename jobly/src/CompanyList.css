.CompanyList-text{
    position: absolute;
    top: 175px;
    left: 50%;
    transform: translate(-50%);
}

.CompanyList-text-loading{
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}



