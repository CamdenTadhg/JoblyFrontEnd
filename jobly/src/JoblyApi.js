import axios from "axios";

// const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";
const BASE_URL = 'http://localhost:3001';

/** API Class.
 *
 * Static class tying together methods used to get/send to to the API.
 * There shouldn't be any frontend-specific stuff here, and there shouldn't
 * be any API-aware stuff elsewhere in the frontend.
 *
 */

class JoblyApi {
  // the token for interaction with the API will be stored here.
  static token;

  static async request(endpoint, data = {}, method = "get") {
    console.log("API Call:", endpoint, data, method);

    //there are multiple ways to pass an authorization token, this is how you pass it in the header.
    //this has been provided to show you another way to pass the token. you are only expected to read this code for this project.
    const url = `${BASE_URL}/${endpoint}`;
    const headers = { authorization: `Bearer ${JoblyApi.token}` };
    const params = (method === "get")
        ? data
        : {};

    try {
      console.log('headers are', headers);
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.response);
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    }
  }

  // Individual API routes
  /** Login a user and receive back a token. LoginData = {username, password}*/
  static async login(loginData){
    try {
      let res = await this.request(`auth/token`, loginData, 'post');
      console.log('res.token is ', res.token);
      JoblyApi.token=res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Signup a user and receive back a token. signupData = {username, password, firstName, lastName, email} */
  static async signup(signupData){
    try{
      let res = await this.request(`auth/register`, signupData, "post");
      JoblyApi.token = res.token;
      return res.token;
    } catch (error){
      return error;
    }
  }

  /** Get list of all companies */
  static async getAllCompanies(){
    let res = await this.request(`companies`);
    return res.companies;
  }

  /** Get a filtered list of companies. searchFilters = {minEmployees, maxEmployees, name} */
  static async getFilteredCompanies(searchFilters) {
    let {minEmployees, maxEmployees, name} = searchFilters;
    let url = 'companies?';
    if (minEmployees) {
      url = url + `minEmployees=${minEmployees}&`;
    }
    if (maxEmployees){
      url = url + `maxEmployees=${maxEmployees}&`;
    }
    if (name) {
      url = url + `name=${name}`
    }
    let res = await this.request(url);
    console.log(res);
    return res.companies;
  }

  /** Get details on a company by handle. */
  static async getCompany(handle) {
    let res = await this.request(`companies/${handle}`);
    console.log(res.company);
    return res.company;
  }

  /**Get list of all jobs */
  static async getAllJobs(){
    let res = await this.request(`jobs`)
    return res.jobs;
  }

  /** Get a filtered list of jobs. searchFilters = {salary, equity, title}*/
  static async getFilteredJobs(searchFilters){
    let {salary, equity, title} = searchFilters;
    let url = 'jobs?'
    if (salary){
      url = url + `minSalary=${salary}&`;
    }
    if (equity) {
      url=url+`hasEquity=${equity}&`;
    }
    if(title) {
      url= url + `title=${title}`;
    }
    let res = await this.request(url);
    return res.jobs;
  }

  /** Get a user's details */
  static async getUserDetails(username){
    let res = await this.request(`users/${username}`);
    return res.user;
  }

  /** Edit a user's details. userData = {firstName, lastName, password, email} */
  static async editUserDetails(username, userData){
    let res = await this.request(`users/${username}`, userData, 'patch');
    console.log('res is', res);
    return res.user;
  }

  /** Delete a user's account */
  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, 'delete');
    return res.deleted;
  }

  /**Create job application for a user */
  static async applyToJob(username, jobId){
    let res = await this.request(`users/${username}/jobs/${jobId}`, {}, "post");
    return res;
  }
}

export default JoblyApi;
