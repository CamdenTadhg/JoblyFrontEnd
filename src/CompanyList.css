.CompanyList-text{
    position: absolute;
    top: 250px;
    left: 50%;
    transform: translate(-50%);
}

.CompanyList-text-loading{
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

.CompanyList-search{
    position: absolute;
    top: 65px;
    left: 50%;
    transform: translate(-50%);
}



