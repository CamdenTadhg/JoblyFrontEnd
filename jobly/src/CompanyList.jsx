import React, {useEffect, useState} from 'react';
import SearchForm from './SearchForm';
import CompanyCard from './CompanyCard';
import JoblyApi from './JoblyApi';
import './CompanyList.css';

function CompanyList(){
    const [isLoading, setIsLoading] = useState(true);
    const [companies, setCompanies] = useState([]);

    //get the list of companies from the API and set state appropriately
    useEffect(() => {
        async function getCompanies() {
            setIsLoading(true);
            let loadCompanies = await JoblyApi.getAllCompanies();
            setCompanies(loadCompanies);
            setIsLoading(false);
        }
        getCompanies();
    }, []);


    if (isLoading) {
        return (
            <div className='CompanyList'>
                <h3 className='CompanyList-text-loading'>Loading &hellip;</h3>
            </div>
        )
    }

    return(
        <div className="CompanyList">
            <div>
                <div className='CompanyList-search'>
                    <SearchForm setData={setCompanies} fields={['name', 'minEmployees', 'maxEmployees']}/>
                </div>
                <div className='CompanyList-text'>
                    {companies.map(company => (
                        <CompanyCard key={company.handle} handle={company.handle} name={company.name} description={company.description} emps={company.numEmployees}/>
                    ))}
                </div>
            </div>
        </div>

    )
}

export default CompanyList;