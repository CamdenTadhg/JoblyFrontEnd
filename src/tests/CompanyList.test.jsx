import React from 'react';
import CompanyList from '../CompanyList';
import {test, expect, afterEach, beforeEach} from 'vitest';
import {render, cleanup, waitFor} from '@testing-library/react';
import UserContext from '../contexts/userContext';
import {MemoryRouter} from 'react-router-dom';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import JoblyApi from '../JoblyApi';

const Wrapper = ({children}) => {
    return(
        <MemoryRouter>
            <UserContext.Provider value='testuser'>
                {children}
            </UserContext.Provider>
        </MemoryRouter>
    )
};

let spyGetAllCompanies;

vi.mock('react-router-dom', async (importOriginal) => {
    const actual = await importOriginal();
    return {
        ...actual,
        Navigate: vi.fn(({to}) => (`Redirected to ${to}`))
    }
});

beforeEach(() => {
    const mock = new MockAdapter(axios);
    mock.onGet('http://localhost:3001/companies').reply(200, {companies: [
        {handle: 'anderson-arias-morrow', name: 'Anderson, Arias, and Morrow', description: 'Somebody program how I. Face give away discussion view act inside. Your official relationship administration here.', numEmployees: 245},
        {handle: 'arnold-berger-townsend', name: 'Arnold, Berger and Townsend', description: "Kind crime at perhaps beat. Enjoy deal purpose serve begin or thought. Congress everything miss tend.", numEmployees: 795}
    ]});
    mock.onGet('http://localhost:3001/companies?').reply(200, {companies: [
        {handle: 'anderson-arias-morrow', name: 'Anderson, Arias, and Morrow', description: 'Somebody program how I. Face give away discussion view act inside. Your official relationship administration here.', numEmployees: 245},
        {handle: 'arnold-berger-townsend', name: 'Arnold, Berger and Townsend', description: "Kind crime at perhaps beat. Enjoy deal purpose serve begin or thought. Congress everything miss tend.", numEmployees: 795}
    ]});
    spyGetAllCompanies = vi.spyOn(JoblyApi, 'getAllCompanies');
});

afterEach(cleanup);

test('renders the CompanyList component', () => {
    render(<CompanyList/>, {wrapper: Wrapper});
});

test('matches the snapshot', async () => {
    const companyList = render(<CompanyList/>, {wrapper: Wrapper});
    await waitFor(() => {
        expect(JoblyApi.getAllCompanies).toHaveBeenCalled();
    });
    expect(companyList).toMatchSnapshot();
});

test('displays the correct content', async () => {
    const {getByText} = render(<CompanyList/>, {wrapper: Wrapper});
    await waitFor(() => {
        expect(JoblyApi.getAllCompanies).toHaveBeenCalled();
    });
    expect(getByText('Anderson, Arias, and Morrow')).toBeInTheDocument();
});

test('displays loading if loading', () => {
    const {getByText} = render(<CompanyList />, {wrapper: Wrapper});
    expect(getByText('Loading', {exact: false})).toBeInTheDocument();
});

test('redirects to home if not logged in', () =>{
    const {getByText} = render(
        <MemoryRouter>
            <CompanyList />
        </MemoryRouter>);
    expect(getByText('Redirected to /')).toBeInTheDocument();
});