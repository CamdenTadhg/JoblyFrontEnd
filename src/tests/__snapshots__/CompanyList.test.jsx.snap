// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="CompanyList"
      >
        <div>
          <div
            class="CompanyList-search"
          >
            <div
              class="SearchForm"
            >
              <form
                class="SearchForm-form"
              >
                <div
                  class="input-item"
                >
                  <label
                    for="name"
                  >
                    Name
                    :
                  </label>
                  <input
                    id="name"
                    name="name"
                    value=""
                  />
                </div>
                <div
                  class="input-item"
                >
                  <label
                    for="minEmployees"
                  >
                    MinEmployees
                    :
                  </label>
                  <input
                    id="minEmployees"
                    name="minEmployees"
                    value=""
                  />
                </div>
                <div
                  class="input-item"
                >
                  <label
                    for="maxEmployees"
                  >
                    MaxEmployees
                    :
                  </label>
                  <input
                    id="maxEmployees"
                    name="maxEmployees"
                    value=""
                  />
                </div>
              </form>
            </div>
          </div>
          <div
            class="CompanyList-text"
          >
            <a
              class="CompanyCard"
              href="/companies/anderson-arias-morrow"
            >
              <div
                class="CompanyCard-card card"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title"
                  >
                    <b>
                      Anderson, Arias, and Morrow
                    </b>
                  </div>
                  <p
                    class="card-text"
                  >
                    Somebody program how I. Face give away discussion view act inside. Your official relationship administration here.
                     
                    <span
                      class="CompanyCard-emps"
                    >
                      (Employees: 
                      245
                      )
                    </span>
                  </p>
                </div>
              </div>
            </a>
            <a
              class="CompanyCard"
              href="/companies/arnold-berger-townsend"
            >
              <div
                class="CompanyCard-card card"
              >
                <div
                  class="card-body"
                >
                  <div
                    class="card-title"
                  >
                    <b>
                      Arnold, Berger and Townsend
                    </b>
                  </div>
                  <p
                    class="card-text"
                  >
                    Kind crime at perhaps beat. Enjoy deal purpose serve begin or thought. Congress everything miss tend.
                     
                    <span
                      class="CompanyCard-emps"
                    >
                      (Employees: 
                      795
                      )
                    </span>
                  </p>
                </div>
              </div>
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="CompanyList"
    >
      <div>
        <div
          class="CompanyList-search"
        >
          <div
            class="SearchForm"
          >
            <form
              class="SearchForm-form"
            >
              <div
                class="input-item"
              >
                <label
                  for="name"
                >
                  Name
                  :
                </label>
                <input
                  id="name"
                  name="name"
                  value=""
                />
              </div>
              <div
                class="input-item"
              >
                <label
                  for="minEmployees"
                >
                  MinEmployees
                  :
                </label>
                <input
                  id="minEmployees"
                  name="minEmployees"
                  value=""
                />
              </div>
              <div
                class="input-item"
              >
                <label
                  for="maxEmployees"
                >
                  MaxEmployees
                  :
                </label>
                <input
                  id="maxEmployees"
                  name="maxEmployees"
                  value=""
                />
              </div>
            </form>
          </div>
        </div>
        <div
          class="CompanyList-text"
        >
          <a
            class="CompanyCard"
            href="/companies/anderson-arias-morrow"
          >
            <div
              class="CompanyCard-card card"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title"
                >
                  <b>
                    Anderson, Arias, and Morrow
                  </b>
                </div>
                <p
                  class="card-text"
                >
                  Somebody program how I. Face give away discussion view act inside. Your official relationship administration here.
                   
                  <span
                    class="CompanyCard-emps"
                  >
                    (Employees: 
                    245
                    )
                  </span>
                </p>
              </div>
            </div>
          </a>
          <a
            class="CompanyCard"
            href="/companies/arnold-berger-townsend"
          >
            <div
              class="CompanyCard-card card"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-title"
                >
                  <b>
                    Arnold, Berger and Townsend
                  </b>
                </div>
                <p
                  class="card-text"
                >
                  Kind crime at perhaps beat. Enjoy deal purpose serve begin or thought. Congress everything miss tend.
                   
                  <span
                    class="CompanyCard-emps"
                  >
                    (Employees: 
                    795
                    )
                  </span>
                </p>
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
